server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java

```diff
@@ -385,26 +386,28 @@ public DiscoveryNodes newNode(DiscoveryNode node) {
      * Returns the changes comparing this nodes to the provided nodes.
      */
     public Delta delta(DiscoveryNodes other) {
-        List<DiscoveryNode> removed = new ArrayList<>();
-        List<DiscoveryNode> added = new ArrayList<>();
+        final List<DiscoveryNode> removed = new ArrayList<>();
+        final List<DiscoveryNode> added = new ArrayList<>();
         for (DiscoveryNode node : other) {
-            if (!this.nodeExists(node)) {
+            if (this.nodeExists(node) == false) {
                 removed.add(node);
             }
         }
         for (DiscoveryNode node : this) {
-            if (!other.nodeExists(node)) {
+            if (other.nodeExists(node) == false) {
                 added.add(node);
             }
         }
+
         DiscoveryNode previousMasterNode = null;
+        if (other.masterNodeId != null) {
+            previousMasterNode = other.getMasterNode();
```

Comment:
`getMasterNode()` already returns null if `masterNodeId` is null.
Maybe cleaner to make that explicit in the `getMasterNode()` method (and not rely on the map implementation to do that for us) and also use `@Nullable` on the return type. The effect is that we won't need these conditions here and can just write `return new Delta(other.getMasterNode(), getMasterNode(), ...`